File: Koko Eating Bananas.cpp
class Solution {
  public:
    
    int count(vector<int>& arr, int mid) {
        
        int count = 0;
        int n = arr.size();
        
        for(int i=0; i<n; i++){
            count = count + (arr[i] / mid);
            
            if(arr[i] % mid > 0){
                count++;
            }
        }
        return count;
    }
    int kokoEat(vector<int>& arr, int k) {
        
        int right = *max_element(arr.begin(),arr.end());
        int left = 1;
        int ans = 0;
        
        while(left <= right){
            int mid = left + (right -left)/2;
            
            if(count(arr,mid) <= k){
                ans = mid;
                right = mid - 1;
            }else{
                left = mid + 1;
            }
        }
        return ans;
    }
};

File: Koko Eating Bananas.java
class Solution {
    public boolean check(int arr[], int mid, int k)
    {
        int temp=0;
        for(int i=0; i<arr.length; i++)
        {
            int a = arr[i]/mid;
            if(arr[i]%mid != 0)
            {
                a++;
            }
            temp = temp + a;
        }
        if(temp <= k)
        {
            return true;
        }
        return false;
    }
    public int kokoEat(int[] arr, int k) 
    {
        int l=1;
        int r= 1000000;
        while(l<=r)
        {
            int mid = l + (r-l)/2;
            boolean res = check(arr, mid, k);
            if(res == false)
            {
                l =mid+1;
            }
            else
            {
                r=mid-1;
            }
        }
        return l;
    }
}
